src/screencast/wlr_screencast.c

```diff
@@ -313,6 +313,205 @@ static void wlr_init_xdg_outputs(struct xdpw_screencast_context *ctx) {
 	}
 }
 
+static bool exec_chooser(char *cmd, bool readin, int p1[2], int p2[2]) {
+	logprint(TRACE, "exec chooser called: cmd %s, readin %d, pipe1 (%d,%d), pipe2 (%d,%d)", cmd, readin, p1[0], p1[1], p2[0], p2[1]);
+	pid_t pid = fork();
+
+	if (pid < 0) {
+		perror("fork");
+		close(p1[0]);
+		close(p2[1]);
+		return false;
+	} else if (pid == 0) {
+		close(p2[0]);
+
+		if (readin) {
+			dup2(p1[0], STDIN_FILENO);
+		}
+		dup2(p2[1], STDOUT_FILENO);
+		close(p1[0]);
+		close(p2[1]);
+
+		int err;
+		err = execl("/bin/sh", "/bin/sh", "-c", cmd, (void *)NULL);
+
+		if (err == -1) {
+			perror("execl");
+		}
+		_exit(127);
+	}
+
+	close(p1[0]);
+	close(p2[1]);
+
+	int status;
+	wait(&status);
+	if (WIFEXITED(status)) {
+		if (WEXITSTATUS(status) == 127) {
+			return false;
+		} else {
+			return true;
+		}
+	}
+	return false;
+}
+
+static bool wlr_output_chooser(struct xdpw_output_chooser *chooser,
+		struct wl_list *output_list, struct xdpw_wlr_output **output) {
+	logprint(DEBUG, "wlroots: output chooser called");
+	struct xdpw_wlr_output *out, *tmp;
+	size_t namelength = 0;
+	char *name = NULL;
+	bool readin = false;
+	FILE *f = NULL;
+
+	int p1[2]; //p -> c
+	int p2[2]; //c -> p
+
+	if (pipe(p1) == -1) {
+		perror("pipe1");
+		logprint(ERROR, "Failed to open pipe1");
+		*output = NULL;
+		return false;
+	}
+	if (pipe(p2) == -1) {
+		perror("pipe2");
+		logprint(ERROR, "Failed to open pipe2");
+		close(p1[0]);
+		close(p1[1]);
+		*output = NULL;
+		return false;
+	}
+
+	if (chooser->type == XDPW_CHOOSER_DMENU) {
+		f = fdopen(p1[1], "w");
+		if (f == 0) {
+			perror("fdopen pipe1");
+			logprint(ERROR, "Failed to create stream writing to pipe1");
+			close(p1[0]);
+			close(p1[1]);
+			close(p2[0]);
+			close(p2[1]);
+			*output = NULL;
+			return false;
+		}
+		wl_list_for_each_safe(out, tmp, output_list, link) {
+			fprintf(f, "%s\n", out->name);
+		}
+		fclose(f);
+		readin = true;
+	} else {
+		close(p1[1]);
+	}
+
+
+	if (exec_chooser(chooser->cmd, readin, p1, p2)) {
+		f = fdopen(p2[0], "r");
+		if (f == 0) {
+			perror("fdopen pipe2");
+			logprint(ERROR, "Failed to create stream reading from pipe2");
+			close(p2[0]);
+			*output = NULL;
+			return true;
+		}
+
+		ssize_t nread = getline(&name, &namelength, f);
+		fclose(f);
+		if (nread < 0) {
+			perror("getline failed");
+			*output = NULL;
+			return true;
+		}
+
+		//Strip newline
+		char *p = strchr(name, '\n');
+		if (p != NULL) {
+			*p = '\0';
+		}
+
+		logprint(TRACE, "wlroots: output chooser %s selects output %s", chooser->cmd, name);
+		wl_list_for_each_safe(out, tmp, output_list, link) {
+			if (strcmp(out->name, name) == 0) {
+				free(name);
+				*output = out;
+				return true;
```

Comment:
We can just `break` here, to de-duplicate the cleanup.

Reply:
check